third_party/move/move-vm/runtime/src/storage/layout_cache.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Data structures for caching layouts of VM values ina long-living, concurrent cache. The design
+//! goal is to make sure that cache hit is semantically equivalent to layout construction. That is,
+//! if layout construction costs 5 units of gas, so does the cache hit. Failed layout construction
+//! never results in insertion into the cache.
+//!
+//! Additionally, caches store only roots of the layout tree. There is NO caching for sub-layouts
+//! because:
+//!   1. This is more error-prone because enforcing of semantic equivalence when reading sub-layout
+//!      is more difficult: e.g., one needs to ensure the depth and size of layouts are correct.
+//!   2. Arguably, we need layouts for root-like values only (e.g., those with `key` ability).
+
+use crate::LayoutWithDelayedFields;
+use move_binary_format::errors::PartialVMResult;
+use move_core_types::language_storage::ModuleId;
+use move_vm_types::{loaded_data::struct_name_indexing::StructNameIndex, ty_interner::TypeVecId};
+use std::collections::BTreeSet;
+use triomphe::Arc as TriompheArc;
+
+/// Set of unique modules that are used to construct a type layout. Iterating over the modules uses
+/// the same order as when constructing layout.
+#[derive(Debug, Default)]
+pub struct DefiningModules {
+    modules: BTreeSet<ModuleId>,
```

Comment:
So we do unconditional flush right now, but we could do better in the future since we track this?

Reply:
Unconditional flush on module upgrade is happening somewhere in the aptos vm, isn't it?

I mean that you only invalidate cached (enum) layout for which some module in the defining set has been upgraded. 

You seem to see this set here mostly for gas charging, but there is more which can be done.

Reply:
It happens in Block-STM, Aptos VM does not know when transaction is committed and is no longer speculative, so we need to do it there.

We can invalidate only published enum, but then we should cache enums (which we want actually), not just top-level resources. We can do this change as a follow-up as it mostly about replacing `MoveTypeLayout` with better representation.